analytic_engine/src/meta/details.rs

```diff
@@ -291,86 +327,231 @@ impl Manifest for ManifestImpl {
+        let path = self.current_snapshot_path();
+        let payload = {
+            let get_res = self.store.get(&path).await;
+            if let Err(object_store::ObjectStoreError::NotFound { path, source }) = &get_res {
+                warn!(
+                    "Current snapshot file doesn't exist, path:{}, err:{}",
+                    path, source
+                );
+                return Ok(None);
+            }
+
+            get_res
+                .context(FetchCurrentSnapshot)?
+                .bytes()
+                .await
+                .context(FetchCurrentSnapshot)?
+        };
+        let current_snapshot = meta_update::CurrentSnapshot::decode(payload.as_bytes())
+            .context(DecodeCurrentSnapshot)?;
+        Ok(Some(current_snapshot.sequence))
+    }
+}
+
+#[async_trait]
+impl MetaUpdateSnapshotStore for ObjectStoreBasedSnapshotStore {
+    /// Store the latest snapshot to the underlying store, and then update the
+    /// `current_snapshot` to ensure the mapping to the latest snapshot.
+    /// The old snapshot will be purged when things done.
+    async fn store(&self, snapshot: &Snapshot) -> Result<()> {
+        let snapshot_pb = meta_update::Snapshot::from(snapshot.clone());
+        let payload = snapshot_pb.encode_to_vec();
+        let path = self.snapshot_path(snapshot.end_seq);
+        self.store
+            .put(&path, payload.into())
+            .await
+            .context(StoreSnapshot)?;
+
+        // Fetch the old snapshot sequence before overwriting it.
+        let old_snapshot_seq = match self.load_current_snapshot_seq().await {
+            Ok(v) => v,
+            Err(e) => {
+                warn!(
+                    "Ignore the failure to load current snapshot for following purge, err:{}",
+                    e
+                );
+                None
+            }
+        };
+
+        // Update the current snapshot.
+        let current_snapshot = meta_update::CurrentSnapshot {
+            sequence: snapshot.end_seq,
+        };
+        let payload = current_snapshot.encode_to_vec();
+        let path = self.current_snapshot_path();
+        self.store
+            .put(&path, payload.into())
+            .await
+            .context(UpdateCurrentSnapshot)?;
+
+        // Purge the old snapshot if any, and ignore the failure.
+        if let Some(v) = old_snapshot_seq {
+            let old_snapshot_path = self.snapshot_path(v);
+            if let Err(e) = self.store.delete(&old_snapshot_path).await {
+                warn!(
+                    "Ignore the failure to purge old snapshot:{}, err:{}",
+                    old_snapshot_path, e
+                );
+            } else {
+                debug!("Succeed in purging old snapshot:{}", old_snapshot_path);
+            }
         }
+
+        Ok(())
     }
+
+    /// Load the `current_snapshot` file from the underlying store, and with the
+    /// mapping info in it load the latest snapshot file then.
```

Comment:
Mapping info

Reply:
I guess it should be `mapping info` rather than `Mapping info` here.